### PR TITLE
Improvements in science display

### PR DESCRIPTION
**Pull request by Gualdimar**
What is done:
* Science overflow is now displayed when picking new tech and is taken into account when calculating the number of turns for research
* Notification with the amount of science from a research agreement
* Checking for tech completion at the beginnig of a turn(for cases when there are 2 turns left before the end of the research and another civilization is completing research of the same technology on the current turn and research costs decrease)

### CODE REVIEW

**Comment by yairm210** [comment]
Pics?

**Comment by Gualdimar** [comment]
<details><summary>Notification</summary>


</details>
<details><summary>Science overflow(8800) is displayed in the pick button for each researchable tech</summary>





</details>

**Comment by Gualdimar** [comment]
I just realized that the value of science from a research agreement is written to one common variable `totalOfScienceDuringRA`. And if a player has two agreements in parallel, then when the first one ends, he will receive part of the science from the second.

But this is work for another PR, which requires a process of non-transient `totalOfScienceDuringRA` deprecation

**Comment by Gualdimar** [comment]
> between beakers "stored" into a specific tech

Science overflow is not stored in any technology until the end of the turn. This is excess science left over from the previous research and can be applied to any of the technologies available for research. It didn't display anywhere that you had this excess science, and it was quite strange to see the next turn the research duration decrease from 12 turns to 6.

> plural, sound better?

It is debatable, since the current implementation does not assume the possibility of several parallel researches.
In addition, it is not recommended to make two agreements in one turn, since the excess science is capped, so you can lose a good share of the benefits.

File: core/src/com/unciv/logic/civilization/managers/TechManager.kt [hunk 1]
```diff
@@ -131,9 +133,12 @@ class TechManager : IsPartOfGameInfoSerialization {
 
     fun remainingScienceToTech(techName: String) = costOfTech(techName) - researchOfTech(techName)
 
-    fun turnsToTech(techName: String) = when {
-        civInfo.stats.statsForNextTurn.science <= 0f -> "∞"
-        else -> max(1, ceil(remainingScienceToTech(techName).toDouble() / civInfo.stats.statsForNextTurn.science).toInt()).toString()
+    fun turnsToTech(techName: String): String {
+        val spareScience = if (canBeResearched(techName)) overflowScience else 0
+        return when {
+            civInfo.stats.statsForNextTurn.science <= 0f -> "∞"
```
[the remark this answers is not in the document]
Comment (Gualdimar) [reply core/src/com/unciv/logic/civilization/managers/TechManager.kt:139]: Hmm, good point. But, if there is a situation where overflow>cost, what should the game do:
* allow to complete the technology on the current turn
* or complete the technology only on the next turn?

**Comment by Gualdimar** [comment]
> see a sum

Currently if you look at the technology, you will see (0/19404), science overflow(8800) is hidden and on the next runt you`ll see (10423/19404) though science income is only 1623 per turn.

**Comment by yairm210** [comment]
In an 'immediate overflow' situation I would complete the tech this turn. This could even happen multiple times, if you have a super high overflow big enough for several small techs...

**Comment by Gualdimar** [comment]
The correct overflow value is now shown for technologies researched by other civilizations.
Now the possibility of completing the research is checked under the following conditions:
* when the research is completed and the next technology is in the queue
* after clicking the Research button on the tech screen
* at the beginnig of a turn

<details><summary>There is a small problem, when there are more than two techs in the queue that can be researched in one turn, the correct number of turns is calculated only for the first one in the queue</summary>


</details>
I need to think about it.
Will be grateful for reviews.

**Comment by yairm210** [comment]
That should be fine

**Review by yairm210 — APPROVED** [review]
Looks mergeable to me, @Gualdimar tell me when you think you're done with this 👍🏿

**Comment by Gualdimar** [comment]
> reorder, fix the broken regions, and document

sounds like another PR to me


> you'll get _two_ techs, not more

yes, overflow is capped in `limitOverflowScience()` according to the rules of the BnW update.

> 'Cuz nobody's ever gonna get that much naturally

An additional check doesn't cost that much, and if more than 2 techs are not possible, it doesn't hurt much, does it? Can mods have an impact here?


> will be tempted to "fix" the double inflation

You're right, I extracted that part to `getScienceModifier()`, DNRY

Thank you for your contribution.
@yairm210, if there are no corrections to the last commit, then you can merge

**Pull request merged**